src/senaite/patient/upgrade/v01_04_000.py

```diff
@@ -135,11 +138,76 @@ def fix_samples_middlename(tool):
 
         obj = api.get_object(brain)
         brain_fullname = brain.getPatientFullName
-        obj_fullname = obj.getPatientFullName()
-        if obj_fullname != brain_fullname:
+        try:
+            obj_fullname = obj.getPatientFullName()
+            reindex = obj_fullname != brain_fullname
+        except AttributeError:
+            # AttributeError: 'unicode' object has no attribute 'get'
+            # Value is not stored as a dict. These cases are handled by 1406
+            reindex = False
+
+        if reindex:
             obj.reindexObject()
 
         # Flush the object from memory
         obj._p_deactivate()
 
     logger.info("Fix samples middle name [DONE]")
+
+
+def fix_samples_without_middlename(tool):
+    """Walks through registered samples and sets an empty middlename to those
+    that do not have a middle name set
+    """
+    logger.info("Fix samples without middle name ...")
+    query = {"portal_type": "AnalysisRequest"}
+    brains = api.search(query, SAMPLE_CATALOG)
+    total = len(brains)
+    for num, brain in enumerate(brains):
+        if num and num % 100 == 0:
+            logger.info("Processed objects: {}/{}".format(num, total))
+
+        if num and num % 1000 == 0:
+            # reduce memory size of the transaction
+            transaction.savepoint()
+
+        obj = api.get_object(brain)
+        field = obj.getField("PatientFullName")
+        value = field.get(obj)
+        if value is None:
+            obj._p_deactivate()
+            continue
+
+        if isinstance(value, six.string_types):
+            # old instance with the fullname stored as str
+            value = {"firstname": value}
+
+        if not isinstance(value, dict):
+            # found some ZPublisher records!!
+            # (Pdb++) type(value)
+            # <class 'ZPublisher.HTTPRequest.record'>
+            try:
+                value = dict(value)
+            except:
+                str_type = repr(type(value))
+                logger.error("Type not supported: {}".format(str_type))
+                obj._p_deactivate()
+                continue
```

Comment:
Maybe here better the explicit approach:
```python
keys = ["firstname", "middlename", "lastname"]
try:
    for k, v in value.items():
        if k not in keys:
            continue
        value[k] = v
except AttributeError:
    pass
```

Reply:
It directly relies now on the field setter